osutil/inotify/inotify_linux.go

```diff
@@ -0,0 +1,319 @@
+//go:build linux
+// +build linux
+
+// Copyright 2010 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+/*
+Package inotify implements a wrapper for the Linux inotify system.
+
+Example:
+    watcher, err := inotify.NewWatcher()
+    if err != nil {
+        log.Fatal(err)
+    }
+    err = watcher.Watch("/tmp")
+    if err != nil {
+        log.Fatal(err)
+    }
+    for {
+        select {
+        case ev := <-watcher.Event:
+            log.Println("event:", ev)
+        case err := <-watcher.Error:
+            log.Println("error:", err)
+        }
+    }
+
+*/
+package inotify
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"syscall"
+	"unsafe"
+)
+
+// NewWatcher creates and returns a new inotify instance using inotify_init(2)
+func NewWatcher() (*Watcher, error) {
+	fd, errno := syscall.InotifyInit1(syscall.IN_CLOEXEC | syscall.IN_NONBLOCK)
+	if fd == -1 {
+		return nil, os.NewSyscallError("inotify_init", errno)
+	}
+	w := &Watcher{
+		fd:         fd,
+		notifyFile: *os.NewFile(uintptr(fd), ""),
+		watches:    make(map[string]*watch),
+		paths:      make(map[int]string),
+		Event:      make(chan *Event),
+		Error:      make(chan error),
+		done:       make(chan bool),
+	}
+
+	go w.readEvents()
+	return w, nil
+}
+
+// Close closes an inotify watcher instance
+// It sends a message to the reader goroutine to quit and removes all watches
+// associated with the inotify instance
+func (w *Watcher) Close() error {
+	if w.isClosed {
+		return nil
+	}
+	w.isClosed = true
+	for path := range w.watches {
+		w.RemoveWatch(path)
+	}
```

Comment:
The point is that the function does more cleans; specifically it also cleans the internal lists. It's true that the garbage collector will also manage that when the object is destroyed... So... keep it for "elegance"? Or remove it?

Reply:
I found that fsnotify has a v0 branch that is maintained and is compatible with Go 1.13, so I decided to migrate into it because it should be easier to maintain.